Provide a way to use field mappng within mapstruct

filiphr opened the pull request:
This is a PR for #557.

The PR is not complete, there are some things which are missing:
- A lot of tests are missing as I was not sure how to do them properly. Should I only update the current tests by adding public fields, or create completely new tests?
- I am also not 100% satisfied with the `afterWriteVariable` and `finishWriteVariable` in the freemarker. Is it possible this to be done via some macro or something similar?
- Do we need to provide a strategy that turns on/off field mapping?

However, I did it so we can have a discussion and see if I am on the right path.

gunnarmorling commented:
HI @filiphr sorry for being slow, just seeing this PR right now. Going to look into it and provide answers to your questions asap. Thanks a lot for working on this!

filiphr commented:
@gunnarmorling no problem. Whenever you have time. Will wait for your input to continue forward

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java:
```diff
@@ -85,13 +87,13 @@
-    private List<ExecutableElement> alternativeTargetAccessors = null;
+    private List<Accessor> allAccessors = null;
+    private List<Accessor> setters = null;
+    private List<Accessor> adders = null;
```
Would an adder not always be a method? I can't see how it could be a field.

filiphr replied:
Yes they will be, they won't be fields. They are filtered from the `allAccessors`. I was thinking that it is better if we have something more generic, and not have the specific implementations referenced in other places

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java:
```diff
@@ -378,21 +380,33 @@ public boolean isAssignableTo(Type other) {
             }
-            getters = Collections.unmodifiableMap( modifiableGetters );
+
+            //TODO decide if we want to have a mechanism for turning on/off direct field mapping
```
I think I'd give preference to calling getters/setters and fall back to field access if they aren't there and the field is accessible for the generated code (either public or protected and living in the same package).

filiphr replied:
It is already like that. For the property read accessors, first I go through the "getter" methods and then through the fields. A field is added only if a "getter" could not be found.

For the property write accessors it is similar. The fields are last checked, they are added at the end of the `getAlternativeTargetAccessors`. I think I am correctly handling the the strategy there. 

Maybe it would be good to add tests for the specific classes that we have

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/util/Executables.java:
```diff
@@ -190,19 +251,29 @@ private static void addEnclosedElementsInHierarchy(Elements elementUtils, List<E
+    private static void addFields(List<Accessor> alreadyCollected, List<VariableElement> variablesToAdd) {
+        List<Accessor> safeToAdd = new ArrayList<Accessor>( variablesToAdd.size() );
+        for ( VariableElement toAdd : variablesToAdd ) {
+            //TODO should we do some check?
```
What kind of check do you have in mind?

filiphr replied:
I was thinking something like with the overriden methods. Whether we should check if a field is already added (shadowed). However, I think it is not needed, as later on the first found field will be used and the rest will be ignored

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/internal/util/accessor/Accessor.java:
```diff
@@ -0,0 +1,46 @@
+
+    Set<Modifier> getModifiers();
+
+    ExecutableElement getExecutable();
```
I reckon that'd return null for a field accessor?

filiphr replied:
Yes it returns `null` for field accessor. I will add documentation to everything once the details are resolved.

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/abstractclass/AbstractDto.java:
```diff
@@ -22,6 +22,10 @@
  */
 public class AbstractDto implements Identifiable {
+
+    //CHECKSTYLE:OFF
```
Does CS complain about public fields? I'd say we should modify the rule set in this case.

gunnarmorling commented:
> A lot of tests are missing as I was not sure how to do them properly. Should I only update the current tests by adding public fields, or create completely new tests?

That's a good question. Maybe make some amendments to key tests from the existing ones? @agudian, @sjaakd any good ideas around it?

> I am also not 100% satisfied with the afterWriteVariable and finishWriteVariable in the freemarker. Is it possible this to be done via some macro or something similar?

I'll leave that to @sjaakd. Ideally the model element knows itself whether to add the braces or not.

> Do we need to provide a strategy that turns on/off field mapping?

Personally I don't think so. I'd use getter/setter access by default and use field access as fallback if possible. In a way people turn it on or off by the way how they model the source and target types.

> However, I did it so we can have a discussion and see if I am on the right path.

+1 That's much appreciated. I definitely think it's going into the right direction. Thanks a lot for tackling this one!

gunnarmorling commented:
Some more questions:

* How does this play together with decorators and `@BeforeMapping`/`@AfterMapping`? Some tests for that would be nice.
* Does it take visibility of fields into account? I.e. we always could access public fields, package private fields only if the generated class lives in the same class as the bean(s) and never private fields.

filiphr commented:
@gunnarmorling Thanks a lot for your feedback. This is an interesting issue to tackle.

> Personally I don't think so. I'd use getter/setter access by default and use field access as fallback if possible. In a way people turn it on or off by the way how they model the source and target types.

You have a good point here. Fields are used as fallback right now, if a method is there it should be used (should write tests though).

> How does this play together with decorators and @BeforeMapping/@AfterMapping? Some tests for that would be nice.

Honestly, I didn't try it. I will add some tests that use those 2 annotations.

> Does it take visibility of fields into account? I.e. we always could access public fields, package private fields only if the generated class lives in the same class as the bean(s) and never private fields.

Currently the fields are filtered with `Executables.isFieldAccessor`. And at this moment only not static public fields are considered, need to improve this to avoid final fields for writing. I think that taking protected and package protected fields into consideration should be tackled in #572

filiphr commented:
@gunnarmorling regarding the decorators. Do you have some test in mind? I am not seeing how the field accessor will change something here.

filiphr commented on processor/src/main/resources/org/mapstruct/ap/internal/model/assignment/ArrayCopyWrapper.ftl:
```diff
@@ -20,11 +20,11 @@
 <#if (thrownTypes?size == 0) >
     <@includeModel object=ext.targetType/> ${localVarName} = <@_assignment/>;
-    ${ext.targetBeanName}.${ext.targetWriteAccessorName}( Arrays.copyOf( ${localVarName}, ${localVarName}.length ) );
+    ${ext.targetBeanName}.${ext.targetWriteAccessorName}${afterWriteVariable}Arrays.copyOf( ${localVarName}, ${localVarName}.length )${finishWriteStatement};
```
@sjaakd These are the points which I don't really like in my PR. The getters are similar.

filiphr commented on processor/src/main/resources/org/mapstruct/ap/internal/model/MethodReference.ftl:
```diff
@@ -37,7 +37,7 @@
                 <@includeModel object=ext.targetType raw=true/>.class<#t>
             <#elseif param.mappingTarget>
-                 ${ext.targetBeanName}.${ext.targetReadAccessorName}()
+                 ${ext.targetBeanName}.${ext.targetReadAccessorName}
```
@sjaakd This is an example getter in the templates

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/util/ValueProvider.java:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @author Filip Hrisafov
+ */
+public class ValueProvider {
```
@sjaakd This is something which I implemented for the getters, so either the getter with () or the field is returned. We need something similar for the assignment

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/assignment/AssignmentWrapper.java:
```diff
@@ -81,8 +82,26 @@ public AssignmentType getType() {
         return decoratedAssignment.isUpdateMethod();
     }
+
+    public String getAfterWriteVariable() {
```
@sjaakd this and the `getFinishWriteStatement()` are used to return the way we need to start/finish an assignment. If you have some ideas I would be glad to hear them. Maybe you have something better in your refactorings.

sjaakd commented:
@filiphr .. I'll take a look tomorrow evening.. thanks...

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/model/common/Type.java:
```diff
@@ -443,28 +457,32 @@ public boolean isAssignableTo(Type other) {
-                    Type targetType = typeFactory.getReturnType(
-                            typeFactory.getMethodType( (DeclaredType) typeMirror, candidate )
-                    );
+                    Type targetType = typeFactory.getReturnType( (DeclaredType) typeMirror, candidate );
```
Hmm.. is the `typeFactory.getMethodType` not required..? Or is this handled by type factory?

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java:
```diff
@@ -296,10 +320,28 @@ public Parameter getSingleParameter(DeclaredType includingType, ExecutableElemen
+    }
+
+    private boolean isExecutableType(TypeMirror accessorType) {
+        return accessorType.getKind() == TypeKind.EXECUTABLE && accessorType instanceof ExecutableType;
```
I reckon the && is not needed, right?

sjaakd replied:
I think you can safely fix this.. :smile:

sjaakd commented:
@filiphr . Nice work!!. I've added a remark on your questions. Have a look if it makes sense. 

Some generals:
* I did not review the tests yet.. I'll do so soon (this weekend).
* As an add on (please different PR) we could (I think that @gunnarmorling mentioned that already) the long lasting discussion on Lombok. E.g. we could add a `@MapperConfiguration` `@Mapper` attribute  `supportLombok` or rather, something more Lombok agnostic as `useFieldsInAbsenceOfAccessors` (contest for better names) or something along those lines. That would inspect fields and make assumptions on getter/setter names which are not yet present when MapStruct is active in the compilation cycle.

filiphr commented:
@sjaakd Thanks for the remarks. I'll try out the CommonMacros.

As for Lombok, yes we could make another PR and find a nice way of reusing this new feature. Currently, only public not static and not final fields are considered as read/write accessors and this is done in the Util functions, maybe we need to move this filtering to somewhere were we have access to the entire Mapping configuration

filiphr commented:
@gunnarmorling, @sjaakd I rebased on our latest master and applied the changes from your comments. @sjaakd Thanks a lot for the `CommonMacros` it helped me a lot and I implemented the assignments much better 😄.

If you have some idea which of our current tests can be extended to work with `public` fields it would be great. There are already tests which I updated, but I am not sure if it is enough

sjaakd commented on processor/src/test/java/org/mapstruct/ap/test/nestedproperties/simple/SimpleMapper.java:
```diff
@@ -31,6 +31,7 @@
     SimpleMapper MAPPER = Mappers.getMapper( SimpleMapper.class );
 
     @Mappings( { @Mapping( target = "longValue", source = "props.longValue" ),
+        @Mapping( target = "publicLongValue", source = "props.publicLongValue" ),
```
do you have a reverse case somewhere as well?

filiphr replied:
Is it enough if I add a reverse method with `@InheritInverseConfiguration` or should I make a completely new one?

sjaakd replied:
keep it simple.. Just reverse (`@InheritInverseConfiguration`) and make sure there are no sneaky issues :).

sjaakd commented:
Hey @filiphr .. Monumental task 👍 .. Only one comment on the reverse nested testing.. This is a tricky area. Perhaps one case to make sure nothing broke there. Otherwise, I'm ok. @gunnarmorling , @agudian WDYT?

filiphr commented:
@sjaakd The reverse mapping works now as well. The only thing that this is missing at the moment is to update the documentation. I'll do this over the weekend. 

Should we log somewhere if a user starts using field mappings?

sjaakd commented:
> Should we log somewhere if a user starts using field mappings?

We should.. But probably not as part of this PR. I think that would demand a new `ReportingPolicy.INFO`, since I think that a warning would be not really applicable. Perhaps open up a new issue?

sjaakd commented on processor/src/main/java/org/mapstruct/ap/internal/util/Executables.java:
```diff
@@ -145,14 +184,35 @@ private static TypeElement asTypeElement(TypeMirror mirror) {
+            if ( accessor.getExecutable() != null ) {
+                executables.add( accessor.getExecutable() );
+            }
+
```
empty line

sjaakd commented:
@filiphr .. I've some very tiny comments left. Have a look.

The only additional thing that sprang to my mind was the `AccessorNamingStrategy`. I'm not sure how relevant that use case is, but it ties read-accessor/write-accessor/presence-checker to a property by naming convention. For the read-accessor/write-accessor there's obviously no issue when calling public fields (properties) directly. However.. are there scenarios conceivable were you would maintain a separate boolean next to the field were you would maintain whether the property is presence or not? And would you do that by some kind of naming convention?

sjaakd commented:
otherwise.. as far as I'm concerned you are ready to squash and go with this one..

agudian commented:
Why should we log something in case the field mapping kicks in?

I find the info-logging in annotation processors always a bit annoying, as they are barely visible anywhere in the IDEs and not at all visible in mvn/javac builds. They just pile up in the Eclipse workspace log files... 😕

sjaakd commented:
I think there is a more general need to log the decisions Mapstruct makes, for debug purposes. I , for instance, find it hard why MapStruct reaches a decision sometimes, and I have an advantage over a non author.

agudian commented:
@sjaakd: #37 should cover that, right?

sjaakd commented:
> @sjaakd: #37 should cover that, right?

Right! Thats what I meant.

filiphr commented:
@agudian Here I mentioned logging on warning level, because this is something new and we can write that in the log message. That way if something comes up it might be faster for the user or us to find what went wrong. Some frameworks have `@Beta` annotations to represent that the feature is in Beta, this of course cannot be used for use as they don't call our methods 😄 

@sjaakd  regarding the `AccessorNamingStrategy`. Using a field `Accessor` has lowest priority and if it is picked it doesn't go through the `AccessorNamingStrategy`. I also wrote a comment about the presence-checkers in #557 and reply from @gunnarmorling 

>Good question, I reckon that it doesn't need to be changed. A field named "foo" really can only represent a property named "foo" and it cannot be a "presence checker" or any of the other method types.

I also think that it doesn't really make sense to have another field which would represent a presence-checker for another field. Currently all the presence-checker are explicitly `ExecutableElementAccessor` -> only methods can be presence-checkers.

I'll fix the comments from @sjaakd and add the documentation

agudian commented:
For new features where we expect that they might not yet work in all of the possibile scenarios, we tried to release them in a beta. For new features where we were not sure if the API or the behavior really fits the bill, we marked those features as *experimental* both in the Javadoc of the annotations (if applicable) and in the reference doc (I just saw that the `@BeforeMapping`/`@AfterMapping` annotations still have that tag... we might want to change that 😉).
That, plus the wording in the release announcment and changelog should suffice, I'd say.

For this feature, are there any questionable decisions that could be changed in the future with impact on what's mapped and what not? If not, I would feel that just releasing it in a beta should be just fine.

filiphr commented:
>For this feature, are there any questionable decisions that could be changed in the future with impact on what's mapped and what not? If not, I would feel that just releasing it in a beta should be just fine.

I don't think so. The only thing that can change that is is the relaxing of the visibility type of the field. Currently only public, not static fields can be used as read accessors, and public, not static and not final as write accessors. However, in my opinion this can be tackled in the same time as #572

sjaakd commented:
So.. am I right to conclude we have consensus on this and we can go ahead and merge? If so: lets move on the the next :smile: ..  I want the large impact ones out of the way to continue refactoring

filiphr commented:
@sjaakd I just thought of one more test that I can add. A public field with getter and setter (I know it makes no sense), but in this case we should use the methods and not the field itself :).

To be faster, I can write the documentation in a separate PR, so we can merge this faster.

sjaakd commented:
> To be faster, I can write the documentation in a separate PR, so we can merge this faster.

That would be ok for me.. Perhaps as part of the documentation you can add an example to our examples? You can refer to that from the documentation.

filiphr commented:
>That would be ok for me.. Perhaps as part of the documentation you can add an example to our examples? You can refer to that from the documentation.

Sure I'll add an example to our examples. When I add the documentation.

My last commit fixes a small bug that I found. If there was a getter then the field was not considered as a write accessor.

In the following class:

```java
public class Source {

    public Integer fieldOnlyWithGetter;

    public Integer getFieldOnlyWithGetter() {
        return fieldOnlyWithGetter;
    }
}
```
The write accessor could not be found. I think that now it should be OK to merge

agudian commented on processor/src/test/java/org/mapstruct/ap/test/fields/FieldsMappingTest.java:
```diff
@@ -18,28 +18,36 @@
 
+    @Before
+    public void setUp() throws Exception {
+        Source.INVOCATIONS.clear();
```
Instead of the invocations class, you could just modify the passed values in the setter/getter methods, eg. add 42 or mulitply with `-1` or so... 😉. Then you can avoid the copy of the `Invocation` class.

filiphr replied:
You are right. That is much better, I didn't think about that. I'll change it and squash everything afterwards

filiphr commented:
Thanks to everyone for the reviews. I removed the changes that @agudian wrote, and once the build is done. I will squash and merge everything into master
